### ISSUE TITLE
Extension and user config catalogues

### ISSUE THREAD

**Issue by Ambrevar** [atlas-engineer/nyxt#966]
In https://github.com/atlas-engineer/nyxt/commit/a30b0653906fec0794e91da0721fa9a072d5dc71 we've removed to outdated EXTENSIONS.org document.

With that we lost the KDEconnect snippet which could be inspiring to KDE users.

Shall we include it in a `contributions` folder?

@jmercouris: What happened to the GitHub wiki?

**Comment by jmercouris** [comment]
It might be, but it probably won't be inspiring to them. I think we should have fewer, higher quality points for information. We can't have both a wiki and a document. If we are to promote something via one of our official channels, we must keep it up to date and be sure it works, or we will have a worse. user experience.

**Comment by jmercouris** [comment]
We can readd the wiki to take the place of the extension exposé. Can we also create pages that are limited in scope?

**Comment by Ambrevar** [comment]
The way I see it is that if we include a document in the repo, it must
be kept up-to-date, which is more work for us, especially if it comes to
stuff that maintainers don't use like KDE.

The benefit if that the doc gets versioned, so the user can always
access the doc the corresponds to the version they are currently using.
Just like the manual.

But is it really important to have such a feature for the rather
volatile extensions?

All in all, I think it's better to keep the volatile information in a
separate spot like the wiki.

**Comment by jmercouris** [comment]
I think we are saying similar things. My questions are just about the implementation of the wiki. Can we make a simple single wiki page titled "Extensions", where people list their extensions? Or maybe a index of Init files that different users have?

**Comment by Ambrevar** [comment]
Both?

We could have one subpage per extension.

**Comment by jmercouris** [comment]
I can see the uitlity of both a extension page, and a directory of init files. Both is OK.

I think a single long page is better for all extensions. Otherwise, too much navigation to find what you are looking for.

**Comment by Ambrevar** [comment]
But if the main page lists all extensions, then navigation is simple,
no?

I was thinking of a main page containing the list of all extensions as
links, which would open to the subpage only containing the code of the
corresponding extension.

**Comment by jmercouris** [comment]
It is simple, but a single page is simpler.

**Comment by jmercouris** [comment]
If the page gets unmanagably long, then we convert to a directory, OK?

**Comment by Ambrevar** [comment]
Let's discuss here questions around community-supported extensions.

For now, we've got a https://github.com/atlas-engineer/nyxt/blob/master/documents/EXTENSIONS.org.

1\. Since it's an Org document, I suggest we format it like this:

```
* [[url][name]]: Description
```

The list will be shorter and just as readable.

2\. What kind of catalogue would we like to have?

- Have a repository of submodules instead?  Drawback: it won't include any description.
- Move it to a wiki instead?  We already discussed of having a wiki to list user configs.  Drawback: it's a separate repo to maintain.
- Only list the extensions in a Quickdist when we get one?  We'd need to extend Quicklisp to support metadata search.

3\. Should extensions use a specific package prefix?  `nx` was proposed.

4\. How do we handle versioning?  We want to give a compatibility guarantee to the user.  How do we deal with unmaintained extensions?  The minimum we must require from every extension is to inform the user of the Nyxt commit/version that was tested.

See #930 for an initial discussion.

**Comment by Ambrevar** [comment]
By the way, why are all our documents uppercase? Shall we downcase them? @jmercouris

**Comment by jmercouris** [comment]
they are uppercase so that they stand out, like README.

**Comment by Ambrevar** [comment]
But should we follow this practice?  The `documents` folder has only
uppercase files, so in the end none of them stands out :p

**Comment by Ambrevar** [comment]
See #533 for the KDE connect extension.
@pranavats Are you interested in maintaining this is a third-party extension?

**Comment by jmercouris** [comment]
YES, I THINK SO. EVEN THOUGH EVERY WORD IN THIS SENTENCE IS UPPERCASE, THEY CERTAINLY DRAW THE EYE.

**Comment by Ambrevar** [comment]
I've worked a little bit on the extension loading process.

The gist of it is that we just load extensions in `init.lisp`, configure each extension in their dedicated file and finally configure the rest of Nyxt itself in a "config.lisp" (arbitrary name).

Here is what my config looks like with the invidious handler extension:

```lisp
;; init.lisp

(defvar *my-request-resource-handlers* '())

(load-after-system :invidious-handler (nyxt-init-file "invidious.lisp"))

(load-after-system :slynk (nyxt-init-file "slynk.lisp"))

(load (nyxt-init-file "config.lisp"))

```

```lisp
;; invidious.lisp

(push invidious-handler:invidious-handler *my-request-resource-handlers*)
```

```lisp
;; config.lisp

(define-configuration web-buffer
  ((request-resource-hook
    (reduce #'hooks:add-hook *my-request-resource-handlers*
            :initial-value %slot-default))))

; rest of my config...
```

`nyxt-init-file` is a helper I've just pushed in 6f77196fe5aa8b91247c01ee3307ac525f6e5869.

What do you think of this workflow?
@jmercouris @aartaka @kssytsrk

**Comment by jmercouris** [comment]
I still think it seems tedious to have a separate file for each configuration, but it is the only way to ensure the symbols exist in the reader...

**Comment by Ambrevar** [comment]
Well, we can also load all extensions in init.lisp, then configure
everything in config.lisp, this way we need only 2 files.

The above example has the benefit of being dynamic: if the extension is
not found, the configuration will still be loaded correctly.

**Comment by jmercouris** [comment]
That's true, it won't fail upon inability to load a particular system, which is really nice.

**Comment by aartaka** [comment]
>  The above example has the benefit of being dynamic: if the extension is not found, the configuration will still be loaded correctly.

But we can just wrap some code snippets into a macro that will check the presence of a package. If the package is found, it runs the code, if not, it just skips and (maybe) makes a `log:warn`.

This seems to be simpler, but haven't we had this already with `load-system`?

**Comment by Ambrevar** [comment]
The problem with executing the code in the body of a macro is that the
package symbols won't be understood by the reader.

This is why we place a package specific config in a separate file.
Does that make sense?

**Comment by kssytsrk** [comment]
> The gist of it is that we just load extensions in init.lisp, configure each extension in their dedicated file and finally configure the rest of Nyxt itself in a "config.lisp" (arbitrary name).
Looks good to me.

**Comment by Ambrevar** [comment]
As @kssytsrk suggested, we should probably have a dedicated place where to load extensions beyond the ASDF default registry.

This is the default:

```lisp
CL-USER> asdf:*default-source-registries*
(ASDF/SOURCE-REGISTRY:ENVIRONMENT-SOURCE-REGISTRY
 ASDF/SOURCE-REGISTRY:USER-SOURCE-REGISTRY
 ASDF/SOURCE-REGISTRY:USER-SOURCE-REGISTRY-DIRECTORY
 ASDF/SOURCE-REGISTRY:DEFAULT-USER-SOURCE-REGISTRY
 ASDF/SOURCE-REGISTRY:SYSTEM-SOURCE-REGISTRY
 ASDF/SOURCE-REGISTRY:SYSTEM-SOURCE-REGISTRY-DIRECTORY
 ASDF/SOURCE-REGISTRY:DEFAULT-SYSTEM-SOURCE-REGISTRY)
```

We can add our tree with

```lisp
CL-USER>  (asdf:initialize-source-registry '(:SOURCE-REGISTRY (:TREE (:HOME ".local/share/nyxt/extensions/"))
 :INHERIT-CONFIGURATION))
```

which can be inspected with

```lisp
CL-USER> asdf:*source-registry-parameter*
(:SOURCE-REGISTRY (:TREE (:HOME "common-lisp-BACKUP/")) :INHERIT-CONFIGURATION)
```

1. Any suggestion on the default location?
2. Shall we remove the default registry to avoid accidental loading of non-Nyxt stuff?
3. Shall we add a helper function to easily load .asd in non-standard locations?
    For instance we could have `load-after-system` support an optional parameter which would be the path to the .asd.

```lisp
(load-after-system :foo "my-config" :asd-file "path/to/foo.asd")
```

Thoughts?

**Comment by kssytsrk** [comment]
My opinion is this:
1. Why not just `~/.config/nyxt/ext/` or something like that? To be honest I don't really understand what would be the perks of making it `~/.local/share/nyxt/extensions`, wouldn't it just confuse the user for a bit (besides, it takes longer to type, so maybe less convenient to clone something there...)? 
2. Yep, I think it would be for the better, can't really think up any reasons why would anyone need to load something from the default registry if there's the ext registry.
3. That would be good.

**Comment by Ambrevar** [comment]
kssytsrk <notifications@github.com> writes:

> 1. Why not just `~/.config/nyxt/ext/` or something like that? To be honest I don't really understand what would be the perks of making it `~/.local/share/nyxt/extensions`, wouldn't it just confuse the user for a bit (besides, it takes longer to type, so maybe less convenient to clone something there...)? 

Well, extensions are not configuration, so I don't think it semantically
belongs to ~/.config ;)

In particular, many users may want to put the whole ~/.config/nyxt under
version control, and then accidentally version the "ext" folder.

It's an easy fix, but it's also easy to not trip the users in the first
place.

Finally, about .local being more typing, I don't think that's too
important because when we start integrating the catalogue into Nyxt it
will automatically store the extensions at the right place, the user
won't have to type anything.

Another option is ~/.cache/nyxt.

> 2. Yep, I think it would be for the better, can't really think up any reasons why would anyone need to load something from the default registry if there's the ext registry.

Good point!

**Comment by Ambrevar** [comment]
Another point: we need to make sure our extension system is compatible
with package manager.

As a test, I suggeset we make a Guix package for the invidious handler
and see if it's automatically picked by Nyxt.

**Comment by kssytsrk** [comment]
> Finally, about .local being more typing, I don't think that's too
important because when we start integrating the catalogue into Nyxt it
will automatically store the extensions at the right place, the user
won't have to type anything.

Oh, I see, makes sense. Didn't know integrating the catalogue into Nyxt was in the plan. Then I'm all for `~/.local/share/nyxt/extensions` (because extensions aren't exactly cache either...)

>As a test, I suggeset we make a Guix package for the invidious handler
and see if it's automatically picked by Nyxt.

Sounds nice, but I don't have any experience with Guix packages so wouldn't be able to help here.

**Comment by Ambrevar** [comment]
kssytsrk <notifications@github.com> writes:

> Sounds nice, but I don't have any experience with Guix packages so wouldn't be able to help here.

I can do it, no worries :)

**Comment by jmercouris** [comment]
We now have the following:

https://discourse.atlas.engineer/c/nyxt/configuration-expose/8

We also have the `~/.local/share/nyxt/extensions` directory coded! Closing this. Feel free to reopen if I missed something!

**Comment by Ambrevar** [comment]
> We also have the `~/.local/share/nyxt/extensions` directory coded!

From my understanding, the question revolved around having a catalogue of community
extensions.  Is `~/.local/share/nyxt/extensions` related to this issue?

**Comment by jmercouris** [comment]
We have the catalogue, it is in the EXTENSIONS.org file.

**Comment by Ambrevar** [comment]
But this file is not accessible from a packaged Nyxt (that is, without
the source).

We could simply include a link or a command in Nyxt to open this link.

**Comment by jmercouris** [comment]
Not yet, I think we will come up with a better solution yet.

**Comment by Ambrevar** [comment]
Reopen this issue then?

**Comment by aadcg** [comment]
Fixed by #3457.

Nyxt shouldn't publicly officially advertise any configurations. On the one hand because we must incite the user to write their own config. On the other hand, those works are not of Nyxt's responsibility. It's ok to mention those on informal exchanges. 

I'm not convinced that it was a good idea to place Nyxt extensions under `~/.local/share/nyxt/extensions`. Emacs for example installs packages under its configuration directory.

@kssytsrk rightly notes

> Why not just ~/.config/nyxt/ext/ or something like that? To be honest I don't really understand what would be the perks of making it ~/.local/share/nyxt/extensions, wouldn't it just confuse the user for a bit (besides, it takes longer to type, so maybe less convenient to clone something there...)?

, while @Ambrevar states that extensions aren't configurations. In my view they are configurations maintained by third-parties.
